File: java/src/main/java/org/example/Solver.java
package org.example;

import java.util.*;

/**
 * 24 点核心求解器，支持高级数学运算
 */
public class Solver {
    public static final double TARGET = 24.0;
    public static final double EPS = 1e-6;
    private static final double DIV_EPS = 1e-12;
    private static final double VALUE_BOUND = 1_000_000.0;
    private static final int MAX_FACTORIAL = 10;
    private static final int MAX_UNARY_DEPTH = 2;
    private static final double EXPONENT_BOUND = 10.0;

    /**
     * 解决特定的数字组合
     * @param numbers 4个数字的数组
     * @return 若能得到靶值结果则返回表达式字符串，否则返回 null
     */
    public static String solveCards(int[] numbers) {
        List<State> initialStates = new ArrayList<>();
        for (int number : numbers) {
            initialStates.add(new State(number, String.valueOf(number), 0));
        }
        return dfs(initialStates, new HashSet<>());
    }

    /**
     * 深度优先搜索，探索所有可能的运算路径
     * @param states 当前状态列表
     * @param visited 已访问的状态集合，用于剪枝
     */
    private static String dfs(List<State> states, Set<List<Long>> visited) {
        List<Long> key = canonicalKey(states);
        // NOTE: 如果该状态组合已访问过，说明处于重复计算路径，直接修剪
        if (!visited.add(key)) {
            return null;
        }

        // 仅剩一个数字时，判断是否接近目标值
        if (states.size() == 1) {
            State state = states.get(0);
            if (close(state.value(), TARGET)) {
                return state.expr();
            }
            return null;
        }

        // 任取两个数字进行组合
        for (int i = 0; i < states.size(); i++) {
            for (int j = i + 1; j < states.size(); j++) {
                List<State> remaining = new ArrayList<>(states.size() - 2);
                for (int k = 0; k < states.size(); k++) {
                    if (k != i && k != j) {
                        remaining.add(states.get(k));
                    }
                }

                List<State> leftVariants = expandUnary(states.get(i));
                List<State> rightVariants = expandUnary(states.get(j));

                // 尝试所有可能的单目与双目运算组合
                for (State left : leftVariants) {
                    for (State right : rightVariants) {
                        List<State> combinedStates = combineStates(left, right);
                        for (State combined : combinedStates) {
                            List<State> variants = expandUnary(combined);
                            for (State variant : variants) {
                                List<State> nextStates = new ArrayList<>(remaining);
                                nextStates.add(variant);
                                String answer = dfs(nextStates, visited);
                                if (answer != null) {
                                    return answer;
                                }
                            }
                        }
                    }
                }
            }
        }
        return null;
    }

    /**
     * 计算状态数组的规范化表示，用作哈希化键
     */
    private static List<Long> canonicalKey(List<State> states) {
        List<Long> buckets = new ArrayList<>(states.size());
        for (State state : states) {
            buckets.add(bucket(state.value()));
        }
        Collections.sort(buckets);
        return buckets;
    }

    /**
     * 对左右两个状态进行所有合法的双目运算
     */
    private static List<State> combineStates(State left, State right) {
        List<State> results = new ArrayList<>();

        pushUnique(results, left.value() + right.value(), "(" + left.expr() + " + " + right.expr() + ")");
        pushUnique(results, left.value() - right.value(), "(" + left.expr() + " - " + right.expr() + ")");
        pushUnique(results, right.value() - left.value(), "(" + right.expr() + " - " + left.expr() + ")");
        pushUnique(results, left.value() * right.value(), "(" + left.expr() + " * " + right.expr() + ")");

        if (Math.abs(right.value()) > DIV_EPS) {
            pushUnique(results, left.value() / right.value(), "(" + left.expr() + " / " + right.expr() + ")");
        }
        if (Math.abs(left.value()) > DIV_EPS) {
            pushUnique(results, right.value() / left.value(), "(" + right.expr() + " / " + left.expr() + ")");
        }

        if (validPower(left.value(), right.value())) {
            pushUnique(results, Math.pow(left.value(), right.value()), "(" + left.expr() + " ^ " + right.expr() + ")");
        }
        if (validPower(right.value(), left.value())) {
            pushUnique(results, Math.pow(right.value(), left.value()), "(" + right.expr() + " ^ " + left.expr() + ")");
        }

        if (validLog(left.value(), right.value())) {
            pushUnique(results, Math.log(left.value()) / Math.log(right.value()), "log_" + right.expr() + "(" + left.expr() + ")");
        }
        if (validLog(right.value(), left.value())) {
            pushUnique(results, Math.log(right.value()) / Math.log(left.value()), "log_" + left.expr() + "(" + right.expr() + ")");
        }

        return results;
    }

    /**
     * 扩展可能的一元操作（例如平方根、阶乘）
     */
    private static List<State> expandUnary(State state) {
        List<State> states = new ArrayList<>();
        states.add(state); // 原值
        
        if (state.unaryDepth() >= MAX_UNARY_DEPTH) {
            return states;
        }

        if (state.value() >= -EPS) {
            double result = state.value() > 0.0 ? Math.sqrt(state.value()) : 0.0;
            pushUniqueWithDepth(states, result, "sqrt(" + state.expr() + ")", state.unaryDepth() + 1);
        }

        Integer integerValue = approxInteger(state.value());
        if (integerValue != null && integerValue <= MAX_FACTORIAL) {
            pushUniqueWithDepth(states, (double) factorial(integerValue), "(" + state.expr() + ")!", state.unaryDepth() + 1);
        }

        return states;
    }

    private static void pushUnique(List<State> states, double value, String expr) {
        pushUniqueWithDepth(states, value, expr, 0);
    }

    private static void pushUniqueWithDepth(List<State> states, double value, String expr, int unaryDepth) {
        if (!validNumber(value)) {
            return;
        }
        // 如果相同数值已经存在，则忽略
        for (State state : states) {
            if (close(state.value(), value)) {
                return;
            }
        }
        states.add(new State(value, expr, unaryDepth));
    }

    private static boolean validNumber(double value) {
        return Double.isFinite(value) && Math.abs(value) <= VALUE_BOUND;
    }

    private static boolean validPower(double base, double exponent) {
        if (Math.abs(exponent) > EXPONENT_BOUND) {
            return false;
        }
        // 如果底数为负数，指数必须为整数，否则会出现复数
        if (base < 0.0 && approxInteger(exponent) == null) {
            return false;
        }
        double result = Math.pow(base, exponent);
        return validNumber(result);
    }

    private static boolean validLog(double argument, double base) {
        return argument > 0.0 && base > 0.0 && !close(base, 1.0);
    }

    /**
     * 获取最接近的非负整数值，如果差异超过容差则返回 null
     */
    private static Integer approxInteger(double value) {
        if (value < -EPS) {
            return null;
        }
        double rounded = (double) Math.round(value);
        if (close(value, rounded) && rounded >= 0.0) {
            return (int) rounded;
        } else {
            return null;
        }
    }

    private static long factorial(int value) {
        long result = 1;
        for (int i = 2; i <= value; i++) {
            result *= i;
        }
        return result;
    }

    /**
     * 将浮点数转换为整型离散桶，用于判重哈希
     */
    private static long bucket(double value) {
        if (value >= 0.0) {
            return (long) (value * 1_000_000.0 + 0.5);
        } else {
            return (long) (value * 1_000_000.0 - 0.5);
        }
    }

    /**
     * 判断两个浮点数是否足够相近
     */
    private static boolean close(double left, double right) {
        return Math.abs(left - right) <= EPS;
    }
}

File: csharp/Solver.cs
namespace Advanced24SolverCSharp;

public static class Solver
{
    public const double Target = 24.0;
    public const double Eps = 1e-6;

    private const double DivEps = 1e-12;
    private const double ValueBound = 1_000_000.0;
    private const int MaxFactorial = 10;
    private const byte MaxUnaryDepth = 2;
    private const double ExponentBound = 10.0;

    public static string? SolveCards(byte[] numbers)
    {
        var states = numbers
            .Select(number => new State(number, number.ToString(), 0))
            .ToList();

        return Dfs(states, new HashSet<string>(StringComparer.Ordinal));
    }

    private static string? Dfs(List<State> states, HashSet<string> visited)
    {
        var key = CanonicalKey(states);
        if (!visited.Add(key))
        {
            return null;
        }

        if (states.Count == 1)
        {
            var state = states[0];
            return Close(state.Value, Target) ? state.Expression : null;
        }

        for (var leftIndex = 0; leftIndex < states.Count; leftIndex++)
        {
            for (var rightIndex = leftIndex + 1; rightIndex < states.Count; rightIndex++)
            {
                var remaining = states
                    .Where((_, index) => index != leftIndex && index != rightIndex)
                    .ToList();

                var leftVariants = ExpandUnary(states[leftIndex]);
                var rightVariants = ExpandUnary(states[rightIndex]);

                foreach (var left in leftVariants)
                {
                    foreach (var right in rightVariants)
                    {
                        foreach (var combined in CombineStates(left, right))
                        {
                            foreach (var variant in ExpandUnary(combined))
                            {
                                var nextStates = new List<State>(remaining.Count + 1);
                                nextStates.AddRange(remaining);
                                nextStates.Add(variant);

                                var answer = Dfs(nextStates, visited);
                                if (answer is not null)
                                {
                                    return answer;
                                }
                            }
                        }
                    }
                }
            }
        }

        return null;
    }

    private static string CanonicalKey(IReadOnlyList<State> states)
    {
        var buckets = states
            .Select(state => Bucket(state.Value))
            .Order()
            .ToArray();

        return string.Join('|', buckets);
    }

    private static List<State> CombineStates(State left, State right)
    {
        var results = new List<State>(10);

        PushUnique(results, left.Value + right.Value, $"({left.Expression} + {right.Expression})");
        PushUnique(results, left.Value - right.Value, $"({left.Expression} - {right.Expression})");
        PushUnique(results, right.Value - left.Value, $"({right.Expression} - {left.Expression})");
        PushUnique(results, left.Value * right.Value, $"({left.Expression} * {right.Expression})");

        if (Math.Abs(right.Value) > DivEps)
        {
            PushUnique(results, left.Value / right.Value, $"({left.Expression} / {right.Expression})");
        }

        if (Math.Abs(left.Value) > DivEps)
        {
            PushUnique(results, right.Value / left.Value, $"({right.Expression} / {left.Expression})");
        }

        if (ValidPower(left.Value, right.Value))
        {
            PushUnique(results, Math.Pow(left.Value, right.Value), $"({left.Expression} ^ {right.Expression})");
        }

        if (ValidPower(right.Value, left.Value))
        {
            PushUnique(results, Math.Pow(right.Value, left.Value), $"({right.Expression} ^ {left.Expression})");
        }

        if (ValidLog(left.Value, right.Value))
        {
            PushUnique(
                results,
                Math.Log(left.Value) / Math.Log(right.Value),
                $"log_{right.Expression}({left.Expression})");
        }

        if (ValidLog(right.Value, left.Value))
        {
            PushUnique(
                results,
                Math.Log(right.Value) / Math.Log(left.Value),
                $"log_{left.Expression}({right.Expression})");
        }

        return results;
    }

    private static List<State> ExpandUnary(State state)
    {
        var states = new List<State> { state };
        if (state.UnaryDepth >= MaxUnaryDepth)
        {
            return states;
        }

        if (state.Value >= -Eps)
        {
            var result = state.Value > 0.0 ? Math.Sqrt(state.Value) : 0.0;
            PushUniqueWithDepth(
                states,
                result,
                $"sqrt({state.Expression})",
                (byte)(state.UnaryDepth + 1));
        }

        var integer = ApproxInteger(state.Value);
        if (integer is not null && integer <= MaxFactorial)
        {
            PushUniqueWithDepth(
                states,
                Factorial(integer.Value),
                $"({state.Expression})!",
                (byte)(state.UnaryDepth + 1));
        }

        return states;
    }

    private static void PushUnique(List<State> states, double value, string expression) =>
        PushUniqueWithDepth(states, value, expression, 0);

    private static void PushUniqueWithDepth(
        List<State> states,
        double value,
        string expression,
        byte unaryDepth)
    {
        if (!ValidNumber(value) || states.Any(state => Close(state.Value, value)))
        {
            return;
        }

        states.Add(new State(value, expression, unaryDepth));
    }

    private static bool ValidNumber(double value) =>
        double.IsFinite(value) && Math.Abs(value) <= ValueBound;

    private static bool ValidPower(double @base, double exponent)
    {
        if (Math.Abs(exponent) > ExponentBound)
        {
            return false;
        }

        if (@base < 0.0 && ApproxInteger(exponent) is null)
        {
            return false;
        }

        return ValidNumber(Math.Pow(@base, exponent));
    }

    private static bool ValidLog(double argument, double @base) =>
        argument > 0.0 && @base > 0.0 && !Close(@base, 1.0);

    private static int? ApproxInteger(double value)
    {
        if (value < -Eps)
        {
            return null;
        }

        var rounded = Math.Round(value);
        return Close(value, rounded) && rounded >= 0.0
            ? (int)rounded
            : null;
    }

    private static long Factorial(int value)
    {
        var result = 1L;
        for (var item = 2; item <= value; item++)
        {
            result *= item;
        }

        return result;
    }

    private static long Bucket(double value) =>
        value >= 0.0
            ? (long)(value * 1_000_000.0 + 0.5)
            : (long)(value * 1_000_000.0 - 0.5);

    private static bool Close(double left, double right) =>
        Math.Abs(left - right) <= Eps;

    private sealed record State(double Value, string Expression, byte UnaryDepth);
}
